File: src/main/java/domain/model/openloopfueling/util/AfrLogUtil.java
package domain.model.openloopfueling.util;

import data.contract.AfrLogFileContract;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AfrLogUtil {
    public static List<Map<String, List<Double>>> findAfrLogs(Map<String, List<Double>> afrLog, double minThrottleAngle, double minRpm, double maxAfr, int minPointsAfr) {
        ArrayList<Map<String, List<Double>>> logList = new ArrayList<>();

        List<Double> throttleAngle = afrLog.get(AfrLogFileContract.TPS_HEADER);
        List<Double> rpm = afrLog.get(AfrLogFileContract.RPM_HEADER);
        List<Double> afr = afrLog.get(AfrLogFileContract.AFR_HEADER);

        for (int i = 0; i < throttleAngle.size(); i++) {

            if (throttleAngle.get(i) >= minThrottleAngle && rpm.get(i) >= minRpm && afr.get(i) < maxAfr) {
                if (LogUtil.isValidLogLength(i, minPointsAfr, minThrottleAngle, throttleAngle)) {
                    int endOfLog = LogUtil.findEndOfLog(i, minThrottleAngle, throttleAngle);
                    logList.add(getAfrLog(i, endOfLog, afrLog));
                    i = endOfLog + 1;
                }
            }
        }

        return logList;
    }

    private static Map<String, List<Double>> getAfrLog(int start, int end, Map<String, List<Double>> afrLog) {
        List<Double> rpm = afrLog.get(AfrLogFileContract.RPM_HEADER);
        List<Double> afr = afrLog.get(AfrLogFileContract.AFR_HEADER);
        List<Double> tps = afrLog.get(AfrLogFileContract.TPS_HEADER);

        Map<String, List<Double>> log = new HashMap<>();
        log.put(AfrLogFileContract.RPM_HEADER, rpm.subList(start, end));
        log.put(AfrLogFileContract.AFR_HEADER, afr.subList(start, end));
        log.put(AfrLogFileContract.TPS_HEADER, tps.subList(start, end));

        return log;
    }
}

File: src/main/kotlin/domain/model/openloopfueling/util/AfrLogUtil.kt
package domain.model.openloopfueling.util

import data.contract.AfrLogFileContract

object AfrLogUtil {
    fun findAfrLogs(afrLog: Map<String, List<Double>>, minThrottleAngle: Double, minRpm: Double, maxAfr: Double, minPointsAfr: Int): List<Map<String, List<Double>>> {
        val logList = mutableListOf<Map<String, List<Double>>>()
        val throttleAngle = afrLog[AfrLogFileContract.TPS_HEADER]!!
        val rpm = afrLog[AfrLogFileContract.RPM_HEADER]!!
        val afr = afrLog[AfrLogFileContract.AFR_HEADER]!!

        var i = 0
        while (i < throttleAngle.size) {
            if (throttleAngle[i] >= minThrottleAngle && rpm[i] >= minRpm && afr[i] < maxAfr) {
                if (LogUtil.isValidLogLength(i, minPointsAfr, minThrottleAngle, throttleAngle)) {
                    val endOfLog = LogUtil.findEndOfLog(i, minThrottleAngle, throttleAngle)
                    logList.add(getAfrLog(i, endOfLog, afrLog))
                    i = endOfLog + 1
                    continue
                }
            }
            i++
        }

        return logList
    }

    private fun getAfrLog(start: Int, end: Int, afrLog: Map<String, List<Double>>): Map<String, List<Double>> {
        return mapOf(
            AfrLogFileContract.RPM_HEADER to afrLog[AfrLogFileContract.RPM_HEADER]!!.subList(start, end),
            AfrLogFileContract.AFR_HEADER to afrLog[AfrLogFileContract.AFR_HEADER]!!.subList(start, end),
            AfrLogFileContract.TPS_HEADER to afrLog[AfrLogFileContract.TPS_HEADER]!!.subList(start, end),
            AfrLogFileContract.BOOST_HEADER to afrLog[AfrLogFileContract.BOOST_HEADER]!!.subList(start, end)
        )
    }
}
